executor/src/constant_evaluator/mod.rs

```diff
@@ -75,6 +69,72 @@ fn generate_values<T: FieldElement>(
     }
 }
 
+struct Symbols<'a, T> {
+    pub analyzed: &'a Analyzed<T>,
+    pub computed_columns: &'a HashMap<&'a str, Vec<T>>,
+}
+
+impl<'a, T: FieldElement> SymbolLookup<'a, T, FixedColumnRef<'a>> for Symbols<'a, T> {
+    fn lookup(&self, name: &str) -> Result<Value<'a, T, FixedColumnRef<'a>>, EvalError> {
+        Ok(
+            if let Some((name, _)) = self.computed_columns.get_key_value(name) {
+                Value::Custom(FixedColumnRef { name })
+            } else if let Some((_, value)) = self.analyzed.definitions.get(&name.to_string()) {
+                match value {
+                    Some(FunctionValueDefinition::Expression(value)) => {
+                        evaluator::evaluate(value, self)?
+                    }
+                    Some(FunctionValueDefinition::Mapping(body)) => (Closure {
+                        parameter_count: 1,
+                        body,
+                        environment: vec![],
+                    })
+                    .into(),
+                    _ => Err(EvalError::Unsupported(
+                        "Cannot evaluate arrays and queries.".to_string(),
+                    ))?,
+                }
+            } else {
+                panic!("Symbol {name} not found.")
+            },
+        )
+    }
+
+    fn eval_function_application(
+        &self,
+        function: FixedColumnRef<'a>,
+        arguments: &[Rc<Value<'a, T, FixedColumnRef<'a>>>],
+    ) -> Result<Value<'a, T, FixedColumnRef<'a>>, EvalError> {
+        if arguments.len() != 1 {
+            Err(EvalError::TypeError(format!(
+                "Expected one argument, but got {}",
+                arguments.len()
+            )))?
+        };
+        let Value::Number(row) = arguments[0].as_ref() else {
+            return Err(EvalError::TypeError(format!(
+                "Expected number but got {}",
+                arguments[0]
+            )));
+        };
+        let data = &self.computed_columns[function.name];
+        Ok(Value::Number(data[row.to_degree() as usize % data.len()]))
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub struct FixedColumnRef<'a> {
```

Comment:
The distinction is that `computed_columns` only contain the columns that we already fully evaluated. So this is another point that should be tested: forward references. I mean with the evaluator this is not such a big deal anymore because it can just continue evaluating the other column instead of re-using the already computed values.

Reply:
It's as I thought. Added two tests.